libs/constrained-generators/src/Constrained/SumList.hs

```diff
@@ -0,0 +1,308 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Constrained.SumList where
+
+import Data.List.NonEmpty (NonEmpty (..))
+import qualified Data.List.NonEmpty as NE
+import System.Random (Random (..))
+import Test.QuickCheck (Gen, choose, shuffle, vectorOf)
+
+-- =======================================================
+-- Helper functions for genSizedList
+
+data Solution t = Yes (NonEmpty [t]) | No [String]
+  deriving (Eq)
+
+instance Show t => Show (Solution t) where
+  show (No xs) = "No" ++ "\n" ++ unlines xs
+  show (Yes xs) = "Yes " ++ show xs
+
+-- | Similar to (concat :: [[x]] -> [x]) for lists
+concatNE :: NonEmpty (NonEmpty t) -> NonEmpty t
+concatNE (x :| []) = x
+concatNE (x :| (y : ys)) = x <> concatNE (y :| ys)
+
+partitionNE :: [Solution t] -> ([NonEmpty [t]], [[String]])
+partitionNE xs = go xs ([], [])
+  where
+    go [] (ys, ns) = (reverse ys, reverse ns)
+    go (Yes y : more) (ys, ns) = go more (y : ys, ns)
+    go (No n : more) (ys, ns) = go more (ys, n : ns)
+
+concatSolution :: Show t => t -> String -> t -> Int -> [Solution t] -> Solution t
+concatSolution smallest pName total count sols = case partitionNE sols of
+  ([], n : _) -> No n -- All No
+  (y : ys, _) -> Yes $ concatNE (y :| ys) -- At least one Yes, and all No's skipped ('ys')
+  ([], []) ->
+    No -- The list is empty
+      [ "The sample in pickAll was empty"
+      , "smallest = " ++ show smallest
+      , "pred = " ++ pName
+      , "total = " ++ show total
+      , "count = " ++ show count
+      ]
+
+newtype Cost = Cost Int deriving (Eq, Show, Num, Ord)
+
+firstYesG ::
+  Monad m => Solution t -> (x -> Cost -> m (Cost, Solution t)) -> [x] -> Cost -> m (Cost, Solution t)
+firstYesG nullSolution f xs c = go xs c
+  where
+    go [] cost = pure (cost, nullSolution)
+    go [x] cost = f x (cost + 1)
+    go (x : more) cost = do
+      ans <- f x (cost + 1)
+      case ans of
+        (cost1, No _) -> go more cost1
+        (cost2, Yes y) -> pure (cost2, Yes y)
+
+noChoices :: Show t => Cost -> String -> t -> t -> Int -> [(t, t)] -> Solution t
+noChoices cost p smallest total count samp =
+  No
+    [ "No legal choice can be found, where"
+    , "  predicate = " ++ p
+    , "  smallest = " ++ show smallest
+    , "  total = " ++ show total
+    , "  count = " ++ show count
+    , "  cost = " ++ show cost
+    , "Small sample of what was explored"
+    , show samp
+    ]
+
+-- =====================================================
+
+{-# SPECIALIZE splitsOf :: Int -> [(Int, Int)] #-}
+
+-- | Given 'count', return a list if pairs, that add to 'count'
+--   splitsOf 6 --> [(1,5),(2,4),(3,3)].
+--   Note we don't return reflections like (5,1) and (4,2),
+--   as they have the same information as (1,5) and (2,4).
+splitsOf :: Integral b => b -> [(b, b)]
+splitsOf count = [(i, j) | i <- [1 .. div count 2], j <- [count - i]]
```

Comment:
It is better to use a let binding, cause it will avoid an extra singleton list construction for every element in the resulting list
```suggestion
splitsOf count = [(i, j) | i <- [1 .. div count 2], let j = count - i]
```